### ISSUE TITLE
How to get results of the  Mahalanobis method?

### ISSUE THREAD

**Issue by songyh15** [deeplearning-wisc/vos#27]
Hi,

Thanks for your public code. The OOD detection benchmarks in your paper are great.
I want to know how to get results of the  Mahalanobis method based on your public code?

Thanks.

**Comment by d12306** [comment]
@songyh15 , hi, sorry for the late reply. 

For Mahalanobis distance, it is somewhat cumbersome.  The procedure you should follow is:

1.  storing the penultimate features as the "inter_feat" here in this position: https://github.com/deeplearning-wisc/vos/blob/725b83c08c7e695a5e68fc7c3b70ab0d92cd7590/detection/inference/rcnn_predictor.py#L134

Also, you need to get the label of the predicted object and append it after the features. this step is used for estimating the class-conditional gaussian models on the training data. The code block you should add is:
```
        inter_feat = box_features.cpu().data
        temp_classes = torch.max(box_cls[:, :-1], 1)[1].view(-1, 1)
        inter_feat = torch.cat((inter_feat, temp_classes.cpu().data), 1)
```

2. take the example of training on voc and test on coco  dataset, perform those evaluations IN ORDER: 
          1)  run the evaluation on the voc_custom_val dataset.
          2)  run the evaluation on the voc_custom_train dataset (used to estimate the Gaussian model in Mahalanobis distance).
          3)  run the evaluation on the coco_ood_val dataset.
3. run the following script on the stored penultimate-layer features.

**Comment by d12306** [comment]
```
import pickle
import torch
import torch.nn.functional as F
import numpy as np
import argparse
import pandas as pd
import seaborn as sns
import matplotlib
matplotlib.use('AGG')
import matplotlib.pyplot as plt
from metric_utils import *
import sklearn
from sklearn import covariance

recall_level_default = 0.95


parser = argparse.ArgumentParser(description='Evaluates an OOD Detector',
                                 formatter_class=argparse.ArgumentDefaultsHelpFormatter)
parser.add_argument('--energy', type=int, default=1, help='noise for Odin')
parser.add_argument('--T', default=1., type=float, help='temperature: energy|Odin')
parser.add_argument('--thres', default=1., type=float)
parser.add_argument('--name', default=1., type=str)
args = parser.parse_args()



concat = lambda x: np.concatenate(x, axis=0)
to_np = lambda x: x.data.cpu().numpy()

def print_results(results, stypes):
    mtypes = ['FPR', 'DTERR', 'AUROC', 'AUIN', 'AUOUT']

    for stype in stypes:
        print(' OOD detection method: ' + stype)
        for mtype in mtypes:
            print(' {mtype:6s}'.format(mtype=mtype), end='')
        print('\n{val:6.2f}'.format(val=100.*results[stype]['FPR']), end='')
        print(' {val:6.2f}'.format(val=100.*results[stype]['DTERR']), end='')
        print(' {val:6.2f}'.format(val=100.*results[stype]['AUROC']), end='')
        print(' {val:6.2f}'.format(val=100.*results[stype]['AUIN']), end='')
        print(' {val:6.2f}\n'.format(val=100.*results[stype]['AUOUT']), end='')
        print('')

# ID data
id_train_data = pickle.load(open('/nobackup-slow/dataset/my_xfdu/VOC-Detection/faster-rcnn/vos/random_seed_0/inference/voc_custom_train/standard_nms/corruption_level_0/probabilistic_scoring_res_odd_0.5686.pkl', 'rb'))
id_train_data = torch.stack(id_train_data['inter_feat']).cpu().data.numpy()

# breakpoint()
id = 0
T = 1
id_score = []
ood_score = []

mean_list = []
covariance_list = []

for index in range(1):
    data_train = id_train_data[:, 0:1024]
    mean_class = np.zeros((20, 1024))
    class_id = id_train_data[:, -1].reshape(-1,1)
    data_train = np.concatenate([data_train, class_id], 1)
    sample_dict = {}

    for i in range(20):
        sample_dict[i] = []
    for data in data_train:
        mean_class[int(data[-1])] += data[:-1]
        sample_dict[int(data[-1])].append(data[:-1])

    for i in range(20):
        mean_class[i] = mean_class[i] / len(sample_dict[i])
        for data in sample_dict[i]:
            data -= mean_class[i]
    mean_class = torch.from_numpy(mean_class)
    group_lasso = covariance.EmpiricalCovariance(assume_centered=False)
    X = 0
    # breakpoint()
    for i in range(20):
        if i == 0:
            X = sample_dict[i]
        else:
            # breakpoint()
            X = np.concatenate([X, sample_dict[i]], 0)

    group_lasso.fit(X)
    temp_precision = group_lasso.precision_
    temp_precision = torch.from_numpy(temp_precision).float()
    precision = temp_precision
    covariance_list.append(precision)
    mean_list.append(mean_class)



all_data_in = pickle.load(open('/nobackup-slow/dataset/my_xfdu/VOC-Detection/faster-rcnn/vos/random_seed_0/inference/voc_custom_val/standard_nms/corruption_level_0/probabilistic_scoring_res_odd_0.5686.pkl','rb'))
all_data_in = torch.stack(all_data_in['inter_feat']).cpu().data.numpy()
all_data_out = pickle.load(open('/nobackup-slow/dataset/my_xfdu//VOC-Detection/faster-rcnn/vos/random_seed_0/inference/coco_ood_val/standard_nms/corruption_level_0/probabilistic_scoring_res_odd_0.5686.pkl','rb'))
all_data_out = torch.stack(all_data_out['inter_feat']).cpu().data.numpy()
print(len(all_data_out))
print(len(all_data_in))
all_data_out = torch.from_numpy(all_data_out)#[:, :-1]
all_data_in = torch.from_numpy(all_data_in)#[:, :-1]



gaussian_score = 0
for i in range(20):
    batch_sample_mean = mean_list[-1][i]
    zero_f = all_data_in - batch_sample_mean
    term_gau = -0.5*torch.mm(torch.mm(zero_f.float(),
                                      covariance_list[-1].float()),
                                        zero_f.float().t()).diag()
    if i == 0:
        gaussian_score = term_gau.view(-1,1)
    else:
        gaussian_score = torch.cat((gaussian_score, term_gau.view(-1,1)), 1)
id_score, _ = torch.max(gaussian_score, dim=1)

gaussian_score = 0
for i in range(20):
    batch_sample_mean = mean_list[-1][i]
    zero_f = all_data_out - batch_sample_mean
    term_gau = -0.5*torch.mm(torch.mm(zero_f.float(),
                                      covariance_list[-1].float()),
                             zero_f.float().t()).diag()
    if i == 0:
        gaussian_score = term_gau.view(-1,1)
    else:
        gaussian_score = torch.cat((gaussian_score, term_gau.view(-1,1)), 1)
ood_score, _ = torch.max(gaussian_score, dim=1)

print(len(id_score))
print(len(ood_score))

measures = get_measures(id_score.cpu().data.numpy(),
                        ood_score.cpu().data.numpy(), plot=False)

print_measures(measures[0], measures[1], measures[2], 'maha')



```

**Comment by songyh15** [comment]
@d12306 Hi, thanks for your reply!

I have run the codes as your description based on your released vanilla model, but I fail to get the results reported in your paper. In my experiments on VOC&COCO,  FPR95 is 96.98, AUROC is 58.41 and AUPR is 95.82.

I beleive that the Mahalanobis method is not supposed to be such poor, but i don't know what is wrong.

BTW, in above codes, "all_data_in" and "all_data_out" should be "all_data_in[:, 0:1024]" and "all_data_out[:, 0:1024]".
And I use the threshold 0.5686 same with above codes, which is the "optimal_score_threshold" for your released vanilla model.

**Comment by d12306** [comment]
@songyh15 , hi, we recently updated the openreview about the Mahalanobis score, the original one reported in the paper is based on the classification logits, we found that in this repo, https://github.com/wetliu/energy_ood, they calculated the score based on the penultimate-layer features. 

Using the same dataset, we obtain a very similar result as yours. If you want double-check, you can calculate the mahalanobis score on the logits, it should be similar to the original paper.
